### ISSUE TITLE
Check a way to automatically extract log information

### ISSUE THREAD

**Issue by uazo** [bromite/bromite#944]
### Is your feature request related to privacy?

No

### Is there a patch available for this feature somewhere?

I don't know, but I'll check

### Describe the solution you would like

Sometimes it is necessary to ask users for more information about issues that do not occur in our test environments.
Since not everyone may be able to extract the technical information necessary to understand the extent of the problem, and, considering that the user is not always able to understand the complexity of development on a browser in such a heterogeneous environment as android, I think it is a priority to expose a specific function that is easily accessible and well documented, so that it is possible for us simple developers to better frame the problem without too much trouble.

### Describe alternatives you have considered

None, for now.

**Comment by csagan5** [comment]
Apps do not have access to logs, not even their own?

**Comment by uazo** [comment]
yes they do and the code is already present in chromium. you just have to organize it.

the part to decide is how the user can send us the files, considering that in my opinion they must be encrypted with a password that he must decide.
if we opened an ad hoc mailbox?

**Comment by csagan5** [comment]
What is the sensitive material in these logs?

This is an open source project, everything happens in the open; if there is privacy-sensitive material the users must remove it before posting it on the issue tracker, where you can already attach files.

You cannot ask users to trust you and send their privacy-sensitive information: it is not right to ask this, by principle. We are educating users to not lose their privacy and that includes not trusting specific individuals or companies with their privacy-sensitive information.

**Comment by uazo** [comment]
> What is the sensitive material in these logs?

yes.

> if there is privacy-sensitive material the users must remove it before posting it on the issue tracker, where you can already attach files.
> You cannot ask users to trust you and send their privacy-sensitive information: it is not right to ask this, by principle. We are educating users to not lose their privacy and that includes not trusting specific individuals or companies with their privacy-sensitive information.

I understand you are right.
then the matter becomes more complicated. something must be provided to help the user obfuscate information that he believes he does not want to disclose.

I have some ideas, but we lack developers. I know that many people read: can anyone help? it doesn't have to be a c++ or android application, a completely client web application would be fine (blazor would be fine) so it can be put on a github page.

**Comment by csagan5** [comment]
I am not sure I understand how a client web application would help here; before calling for contributors, what is the idea?

> something must be provided to help the user obfuscate information that he believes he does not want to disclose.

Most users do not want to go to that process of reviewing the privacy-sensitive information before uploading anything. The best would be to get the only logs without privacy-sensitive information.

**Comment by uazo** [comment]
the following code is already present in the codebase:

## ChromeFeedbackCollector
Currently not callable, but can be done via HelpAndFeedbackLauncherImpl.
The reporter (FeedbackReporter) is currently a pure virtual method

information it could collect for us:
* **HistogramFeedbackSource**
Grabs feedback about the UMA histograms
* **LowEndDeviceFeedbackSource**
Provides data about whether the feedback comes from a low-end device.
* **SystemInfoFeedbackSource**
Grabs feedback about the current system
* **VariationsFeedbackSource**
Grabs feedback about the current variations state.

other info useless for us
* UrlFeedbackSource
Grabs feedback about the current URL
* DataReductionProxyFeedbackSource
Grabs feedback about the data reduction proxy.
* IMEFeedbackSource
Grabs feedback about the default IME input method
* PermissionFeedbackSource
Grabs a list of all granted permissions
* StaticScreenshotSource 
A implementation of ScreenshotSource that returns back a Bitmap given to it
* InterestFeedFeedbackSource
Provides information about whether the interest feed is enabled for use in feedback reports.

## LogcatExtractionRunnable
From code:
```
 * Extracts the recent logcat output from an Android device, elides PII sensitive info from it,
 * prepends the logcat data to the caller-provided minidump file, and initiates upload for the crash
 * report.
 *
 * Elided information includes: Emails, IP address, MAC address, URL/domains as well as Javascript
 * console messages.
```
Here, too, the MinidumpUploadServiceImpl must be implemented, to save to a folder accessible to the user
 
## JavaExceptionReporter (and base/android/java_exception_reporter.cc)
```
 * This UncaughtExceptionHandler will create a breakpad minidump when there is an uncaught
 * exception.
 *
 * The exception's stack trace will be added to the minidump's data. This allows java-only crashes
 * to be reported in the same way as other native crashes.
```

the idea is to add an item in the "developer" menu that allows the selection of information to extract to file.
since there will most likely be private information, the idea may be that of provide a client-only web application that rereads that data and allows the user to delete what he wants not to be published.

**Comment by csagan5** [comment]
> * Grabs feedback about the current variations state.

There are no variations in Bromite.

> * Grabs a list of all granted permissions

This can be useful to debug sandbox-related permission issues.

> the idea may be that of provide a client-only web application that rereads that data and allows the user to delete what he wants not to be published.

The code mentions that is already removing PII

**Comment by csagan5** [comment]
> I personally would like a features freeze until we have a reasonable certainty of having good troubleshooting

@uazo I don't think that crowdsourcing testing is the solution here; sure we can use the code that exists and you found here for the general case, but I do not see developing a client-side app for the logs as something on topic for Bromite

We can do a lot more with manual or automated UI tests.

**Comment by uazo** [comment]
> I do not see developing a client-side app for the logs as something on topic for Bromite

sure, I understand. my fault, sometimes I know I exaggerate, sorry. let's go on without that.

**Comment by uazo** [comment]
> We could publish the .mapping files as part of the releases.

Perfect! Thanks.

**Comment by uazo** [comment]
this is an example of what can be extracted with a few code modifications.

[041b2545-953f-4aa5-95d3-e636dc26dc31.forced21442.zip](https://github.com/bromite/bromite/files/6364952/041b2545-953f-4aa5-95d3-e636dc26dc31.forced21442.zip)

does anyone know how to retrieve the file inside "upload_file_minidump" with some shell line?

**Comment by csagan5** [comment]
Where is `upload_file_minidump`? It's not in that zip file you attached.

**Comment by uazo** [comment]
is the result of https://github.com/bromite/bromite/pull/1105

I don't really know what it is either, it's not directly the tombstone, but it should be read with `minidump_stackwalk` and `crsym`

**Comment by uazo** [comment]
> Where is upload_file_minidump ?

it's inside the only zip file, like so:

```
-----MultipartBoundary-SRM0KIoIVeP8CUATHZPIUQkANgwanADl---
Content-Disposition: form-data; name="upload_file_minidump"; filename="041b2545-953f-4aa5-95d3-e636dc26dc31.dmp"
Content-Type: application/octet-stream
```

to extract

```
sudo apt install mpack
munpack -f <file>
```

**Comment by csagan5** [comment]
Yes, now I see what you mean. That file is a MIME archive, it works like email attachments.

**Comment by uazo** [comment]
Uhm.. doesn't work

```
~/chromium/src/out/x86/minidump_stackwalk d0f557a0-ef09-41d3-94b3-f852270b6c7a.dmp
2021-04-24 07:37:27: minidump.cc:5199: ERROR: Minidump header signature mismatch: (0x444d0a0d, 0xd0a4d44) != 0x504d444d
2021-04-24 07:37:27: minidump_stackwalk.cc:101: ERROR: Minidump d0f557a0-ef09-41d3-94b3-f852270b6c7a.dmp could not be read
```

**Comment by csagan5** [comment]
> Uhm.. doesn't work

The MIME archive works fine, as you were able to extract the files within it. What does not work I guess from your pasted output is the minidump decoding. See https://sites.google.com/a/chromium.org/dev/chromium-os/packages/crash-reporting/debugging-a-minidump

**Comment by uazo** [comment]
maybe `munpack` is wrong (I don't know how to use it), the first 4 bytes should be MDMP

**Comment by uazo** [comment]
I tried to extract it by hand and it works.

after generating

```
components/crash/content/tools/generate_breakpad_symbols.py --build-dir=out/x86 --symbols-dir=/tmp/my_symbols/ --binary=out/x86/lib.unstripped/libchrome.so --clear --verbose
```

(1.1gb in /tmp/my_symbols/)

I run

```
~/chromium/src/out/x86/minidump_stackwalk ~/tmp/t.dmp /tmp/my_symbols
```

result: [final.txt](https://github.com/bromite/bromite/files/6369612/final.txt)

**Comment by csagan5** [comment]
> I tried to extract it by hand

How do you extract by hand?

**Comment by uazo** [comment]
> How do you extract by hand?

Simply removing headers, to understand, something that awk or similar can easily do

**Comment by csagan5** [comment]
Fixed in `91.0.4472.143`.
